prog/dns_zone/dns_zone_nexus.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+require "forwardable"
+
+require_relative "../../lib/util"
+
+class Prog::DnsZone::DnsZoneNexus < Prog::Base
+  subject_is :dns_zone
+
+  label def wait
+    dns_zone.dns_servers.each do |dns_server|
+      records_to_rectify = dns_zone.records_dataset
+        .left_join(:seen_dns_records_by_dns_servers, dns_record_id: :id, dns_server_id: dns_server.id)
+        .where(Sequel[:seen_dns_records_by_dns_servers][:dns_record_id] => nil).all
+
+      next if records_to_rectify.empty?
+
+      commands = ["zone-abort #{dns_zone.name}", "zone-begin #{dns_zone.name}"]
+      records_to_rectify.each do |r|
+        commands << if r.tombstoned
+          "zone-unset #{dns_zone.name} #{r.name} #{r.ttl} #{r.type} #{r.data}"
+        else
+          "zone-set #{dns_zone.name} #{r.name} #{r.ttl} #{r.type} #{r.data}"
+        end
+      end
+      commands << "zone-commit #{dns_zone.name}"
+
+      dns_server.vms.each do |vm|
+        outputs = vm.sshable.cmd("sudo -u knot knotc", stdin: commands.join("\n")).split("\n")
+
+        # Passing multiple commands to knotc via stdin is faster compared to running each
+        # command one by one. However, this approach has one drawback; in stdin mode knotc
+        # always with 0 exit code (i.e. no errors would be raised). At least, errors are
+        # being written to stdout, so we can search them manually to see if we need to
+        # raise any errors.
+        outputs.each_with_index do |output, index|
+          next if output == "OK"
+          next if index == 0 && output.include?("no active transaction")
+          next if commands[index].include?("zone-set") && output.include?("such record already exists in zone")
+          next if commands[index].include?("zone-unset") && output.include?("no such record in zone found")
+
+          raise "Rectify failed for #{dns_zone} on #{dns_server}. Command: #{commands[index]}. Output: #{output}"
+        end
+      end
```

Comment:
Considering we are in the context of dns_zone, we have multiple dns_servers that have direct relationship with out context, looping in once more into the VMs here make me uncomfortable. Is there a way to abstract this dns_server.vms.cmd in directly dns_server.cmd?

Reply:
I recall writing a 'runs_on_all_replicas' helper function once 😿